L58-python-async-api.md

```diff
@@ -0,0 +1,1710 @@
+like interface in Golang and Java, except the error is generated in runtime
+instead of compile time.
+
+If the gRPC Python has multiple implementation for a single interface, the use
+of the design pattern provides productivity in unifying their behavior. However,
+almost non interfaces has second implementation, even if they do, they are
+depends directly on our concrete implementation, which should better using
+inheritance or composition than interfaces.
+
+Also, in the past, dependents of gRPC Python have observed several failure
+caused by the interface. The interface constraints our ability to add
+experimental API. Once we change even slightly with interfaces, the downstream
+implementations are likely to break.
+
+Since this is a new opportunity for us to re-design, we need to think cautiously
+about how do we empower our users to extend our classes. For majority of cases,
+we are providing the only implementation for the interface. We should
+convert them into concrete classes.
+
+On the other hand, there are actually one valid use case that we should keep
+abstract class -- interceptors. To be more specific, the following interfaces
+won't be replaced by concrete classes:
+
+* `grpc.ServerInterceptor`
+* `grpc.UnaryUnaryClientInterceptor`
+* `grpc.UnaryStreamClientInterceptor`
+* `grpc.StreamUnaryClientInterceptor`
+* `grpc.StreamStreamClientInterceptor`
+
+## Future Features
+
+### Implicit Context Propagation By `contextvars`
+
+`contextvars` is a Python 3.7 feature that allows applications to set coroutine
+local variables (just like thread local variables). In the past, gRPC Python
+uses thread local to achieve two goals:
+
+* Support distributed tracing library that work across languages;
+* Support deadline propagation along the chain of RPCs.
+
+For distributed tracing library (like OpenCensus), the tracing information is
+preserved as one of the metadata, and it is expected to be implicitly promoted
+from inbound metadata to outbound metadata. The tracing information will be used
+to monitor the life cycle of an request, and the length it has stayed in each
+services.
+
+For deadline propagation, the deadline of upstream server will be implicitly
+pass down to downstream server, so downstream services can react to that
+information and save computation resources or perform flow control.
+
+Acceptence critiria of this feature:
+* The implimentation of such feature should supported by official package, and
+users are not expected to directly access those metadata;
+* Application logic has higher priority than the implicit propagation (e.g.
+  setting timeout explicitly will override the propagated value);
+* Application has the choice to diverse coroutine local variables;
+* The exception error string should be informative (e.g. pointing out the
+  timeout is due to upstream deadline).
+
+Further discussion around this topic, please see related sections under [Rationale].
```

Comment:
I was expecting a first proposal about how this could be implemented :)

Reply:
@gnossen Can you take a look?

Reply:
@pfreixes I have an in-progress example of an implicit context API as a branch off of https://github.com/lidizheng/grpc-api-examples. I'll prioritize getting that up. Do you think that is an appropriate place to continue this discussion. Also, based on your comment above, is it implementation that you're wondering about, or is this still an API-driven discussion?

Reply:
Not in a rush, let's have the discussion in your PR for the `grpc-api-examples` project.